### ISSUE TITLE
`learning_rate_scheduler_type: inv-sqrt-decay` not reaching `initial_learning_rate`

### ISSUE THREAD

**Issue by SamuelLarkin** [awslabs/sockeye#1095]
Hi,
  I'm running experiments and noticed something.  I'm using
```yaml
checkpoint_interval: 1000
learning_rate_scheduler_type: inv-sqrt-decay
learning_rate_warmup: 4000
update_interval: 10
```
with varying `initial_learning_rate`.

I was expecting the learning rate to reach `initial_learning_rate` at step 4 (`4 x checkpoint_interval == learning_rate_warmup`) but it is always way below the expect value.

```sh
for e in */*/*/model; do
  echo $e;
  grep 'initial_learning_rate\|warm' $e/args.yaml;
  sed -n 4p $e/metrics | awk '{print $9}';
  echo;
done
```
```
Committees/en2fr/baseline.lr=0.006325/model
initial_learning_rate: 0.006325
learning_rate_warmup: 4000
learning-rate=0.000100007031002825

Committees/en2fr/baseline/model
initial_learning_rate: 0.06325
learning_rate_warmup: 4000
learning-rate=0.00100007031002825

Committees/fr2en/baseline.lr=0.006325/model
initial_learning_rate: 0.006325
learning_rate_warmup: 4000
learning-rate=0.000100007031002825

Committees/fr2en/baseline/model
initial_learning_rate: 0.06325
learning_rate_warmup: 4000
learning-rate=0.00100007031002825

Debates/en2fr/baseline.lr=0.006325/model
initial_learning_rate: 0.006325
learning_rate_warmup: 4000
learning-rate=0.000100007031002825

Debates/en2fr/baseline/model
initial_learning_rate: 0.06325
learning_rate_warmup: 4000
learning-rate=0.00100007031002825

Debates/fr2en/baseline.lr=0.006325/model
initial_learning_rate: 0.006325
learning_rate_warmup: 4000
learning-rate=0.000100007031002825

Debates/fr2en/baseline/model
initial_learning_rate: 0.06325
learning_rate_warmup: 4000
learning-rate=0.00100007031002825
```

# Metrics
Using `initial_learning_rate: 0.06325` and looking at the `metrics` file, we see that the peak is at the 4th step and that none of the checkpoint's learning rate come close to `initial_learning_rate`.
```sh
cut -f 1,9 model/metrics
```
```
1       learning-rate=0.0002500175775070625
2       learning-rate=0.000500035155014125
3       learning-rate=0.0007500527325211874
4       learning-rate=0.00100007031002825
5       learning-rate=0.0008944900782009826
6       learning-rate=0.000816553988825397
7       learning-rate=0.0007559820954040039
8       learning-rate=0.0007071564978843085
9       learning-rate=0.0006667135400188333
10      learning-rate=0.0006325
11      learning-rate=0.0006030650876978371
12      learning-rate=0.0005773908627033626
13      learning-rate=0.000554739197211696
14      learning-rate=0.0005345600661157868
15      learning-rate=0.0005164340874367867
16      learning-rate=0.000500035155014125
17      learning-rate=0.00048510535544596183
18      learning-rate=0.0004714376652562057
19      learning-rate=0.0004588637281945823
20      learning-rate=0.0004472450391004913
21      learning-rate=0.00043646646628403904
22      learning-rate=0.00042643141300800064
23      learning-rate=0.00041705814942283526
24      learning-rate=0.0004082769944126985
25      learning-rate=0.00040002812401129995
26      learning-rate=0.00039225984813837176
27      learning-rate=0.0003849272418022417
28      learning-rate=0.00037799104770200196
29      learning-rate=0.0003714167888430499
30      learning-rate=0.0003651740452624383
31      learning-rate=0.0003592358601612685
32      learning-rate=0.0003535782489421542
33      learning-rate=0.00034817979072121155
```

# Version
I'm using `Sockeye: 3.1.31, commit 13c63be5e6999102cd8f76065dab618667d54c8d` with `PyTorch: 1.13.1`.

**Comment by SamuelLarkin** [comment]
I just noticed in my previous message's Metrics section that, oddly, we see `initial_learning_rate / 100` at the 10th step.

**Comment by mjdenkowski** [comment]
Hi Samuel,

In Sockeye's [implementation](https://github.com/awslabs/sockeye/blob/main/sockeye/lr_scheduler.py#L123-L129) of inverse square root learning rate decay, the specified learning rate plugs into the [transformer paper](https://arxiv.org/pdf/1706.03762.pdf) formula in place of `d_model**-0.5`. With a learning rate of `0.006325` and `4000` warmup steps, we get `0.006325 * min(4000**-0.5, 4000 * 4000**-1.5) = 0.00010000703100282499` at step `4000`.

Best,
Michael

**Comment by SamuelLarkin** [comment]
I should have reread the paper. 😳 😓 

Thanks for your answer.

**Comment by SamuelLarkin** [comment]
I should have reread the paper. 😳 😓

Thanks for your answer.
